### PR TITLE
Fix for nil ns_prefix caused by upgrade to recursive-open-struct 1.0

### PR DESCRIPTION
**Pull request by bkoski**
Not sure if this is caused by ruby 2.3.0, recursive_open_struct 1.0.0, or some combination of the two but – when trying to create services with kubeclient 1.0.0 on ruby 2.3.0, was getting 

```
405 The server does not allow this method on the requested resource
```

Did some debugging and found that `ns_prefix` in `create_resource` was nil, because keys under `entity_config.metadata['table']` were strings rather than symbols.   So

```
entity_config.metadata['table'][:namespace]
```

is nil, but these expressions work:

```
entity_config.metadata['table']['namespace']
entity_config.metadata.namesapce
```

Not sure why this reference was going through `.metadata['table']`, but as far as I understand ROS 1.0.0 you should always be able to hit `entity_config.metadata.namespace`?

Changed this reference in this PR.  In addition, removed the workaround for aetherknight/recursive-open-struct#9 since 1.0.0 solves for this.

Tests pass; this also fixes the 405 error.

### CODE REVIEW

File: lib/kubeclient/common.rb [hunk 1]
```diff
@@ -223,13 +219,10 @@ def create_entity(entity_type, entity_config, klass)
 
     def update_entity(entity_type, entity_config)
       name = entity_config.metadata.name
-      # to_hash should be called because of issue #9 in recursive open
-      # struct
-      hash = entity_config.to_hash
-      ns_prefix = build_namespace_prefix(entity_config.metadata['table'][:namespace])
+      ns_prefix = build_namespace_prefix(entity_config.metadata.namespace)
```
Comment (simon3z) [inline lib/kubeclient/common.rb]: @bkoski please don't remove the access to `namespace` using `['table'][:namespace]`, it's going to create a regression in applications where `.namespace` is already overridden.

Was that required for 2.3.0?

Comment (simon3z) [inline lib/kubeclient/common.rb]: @bkoski let's use `entity_config.metadata['table']['namespace']` if that works.

**Comment by simon3z** [comment]
@abonas @bkoski why the tests passed in the first place? If we don't have any test covering this part we should probably add them (with this PR possibly).

File: lib/kubeclient/common.rb [hunk 2]
```diff
@@ -202,20 +202,16 @@ def delete_entity(entity_type, name, namespace = nil)
...
 
       # TODO: temporary solution to add "kind" and apiVersion to request
       # until this issue is solved
       # https://github.com/GoogleCloudPlatform/kubernetes/issues/6439
-      hash['kind'] = entity_type
-      hash['apiVersion'] = @api_version
+      entity_config['kind'] = entity_type
+      entity_config['apiVersion'] = @api_version
```
Comment (simon3z) [inline lib/kubeclient/common.rb]: @bkoski you haven't mentioned this change in behavior: you're now changing the original object passed in (`entity_config`).

Comment (abonas) [inline lib/kubeclient/common.rb]: @bkoski I also wonder why is this change needed for this PR, and why the original object is changed (entity_config)

**Comment by abonas** [comment]
> @abonas @bkoski why the tests passed in the first place? If we don't have any test covering this part we should probably add them (with this PR possibly).

I'm puzzled about the tests passing as well

**Comment by abonas** [comment]
@bkoski could you perhaps share a code example of how you were constructing a service?

**Comment by bkoski** [comment]
Did some digging, the error shows up when you assign `.metadata` using a Hash with string keys (rather than symbols, or directly assigning the properties through deep accessors).

In my particular case, I was setting up the service instance after parsing a JSON payload:

```
payload = JSON.parse(params[:service])
service = Kubeclient::Service.new
service.metadata = service_data['metadata']
service.spec     = service_data['spec']
```

It seems the error was caused not by Ruby 2.3, but instead by the ROS 1.0 upgrade.  In ROS 1.0, [all keys are cast to symbols internally](https://github.com/aetherknight/recursive-open-struct/commit/c5c00f5bf0f690d6f8b236651151b3335005cc71#diff-04c6e90faac2675aa89e2176d2eec7d8R25).

Updated the PR to:
  * include a test that exposes the error
  * go back to the original setup, where `entity_config` is converted to a hash to avoid mutating the original param
  * access the namespace as `hash[:metadata][:namespace]`, which appears to work no matter how you initialize the service

**Comment by abonas** [comment]
@bkoski please squash the commits

**Comment by bkoski** [comment]
Just squashed, and added a couple more tests to exercise creation and updates w/ string/symbol keys.

File: lib/kubeclient/common.rb [hunk 3]
```diff
@@ -207,34 +207,31 @@ def delete_entity(entity_type, name, namespace = nil)
...
       # until this issue is solved
       # https://github.com/GoogleCloudPlatform/kubernetes/issues/6439
-      hash['kind'] = entity_type
-      hash['apiVersion'] = @api_version
+      hash[:kind]       = entity_type
+      hash[:apiVersion] = @api_version
       response = handle_exception do
         rest_client[ns_prefix + resource_name(entity_type)]
-        .post(hash.to_json, @headers)
+        .post(entity_config.to_json, @headers)
```
Comment (abonas) [inline lib/kubeclient/common.rb]: shouldn't it be hash that is sent to the server? 
the entity_config doesn't contain kind and apiVersion set on lines 219 and 220

**Comment by bkoski** [comment]
Good call, fixed this and added some some tests that:
  - check payload sent to kubernetes api to make sure it's valid JSON with complete data
  - assert that original entity_confg is not modified

File: lib/kubeclient/common.rb [hunk 4]
```diff
@@ -227,14 +227,11 @@ def create_entity(entity_type, entity_config, klass)
     end
 
     def update_entity(entity_type, entity_config)
-      name = entity_config.metadata.name
-      # to_hash should be called because of issue #9 in recursive open
-      # struct
-      hash = entity_config.to_hash
-      ns_prefix = build_namespace_prefix(entity_config.metadata['table'][:namespace])
+      name      = entity_config[:metadata][:name]
```
Comment (abonas) [inline lib/kubeclient/common.rb]: did the access with metadata.name stopped working?

File: test/test_service.rb [hunk 5]
```diff
@@ -33,6 +33,81 @@ def test_construct_our_own_service
...
+
+    # Check that original entity_config is not modified by kind/apiVersion patches:
+    assert_equal(our_service.kind, nil)
+
+    assert_requested(:post, expected_url, times: 1) do |req|
+      data = JSON.parse(req.body)
+      data['kind'] == 'Service' &&
+        data['apiVersion'] == 'v1' &&
+        data['metadata']['name'] == 'guestbook' &&
+        data['metadata']['namespace'] == 'staging'
```
Comment (abonas) [inline test/test_service.rb]: :+1:

**Comment by abonas** [comment]
can you please change the PR title to a more relevant one? the mention of ruby 2.3 is misleading as the issue is eventually not related to that.

File: lib/kubeclient/common.rb (hunk 4 above)
Comment (bkoski) [inline lib/kubeclient/common.rb]: `metadata.name` was working, but my thought was that all references into `entity_config` should be consistent.  In the original version of the PR the reference to ns_prefix was
```
build_namespace_prefix(entity_config.metadata.namespace)
```
but seemed like [the preference](https://github.com/abonas/kubeclient/pull/150#discussion_r49056553) was towards using `[]` rather than method references.

**Comment by bkoski** [comment]
Just updated title.

File: lib/kubeclient/common.rb (hunk 4 above)
Comment (abonas) [inline lib/kubeclient/common.rb]: > but seemed like the preference was towards using [] rather than method references

yes, because we are aware with override specific issues of the 'namespace' part :)
but it's ok.

File: test/test_service.rb [hunk 6]
```diff
@@ -111,26 +186,49 @@ def test_get_service
...
+    service.metadata = {}
+    service.metadata.name      = name
+    service.metadata.namespace = 'development'
+
+    expected_url = "http://localhost:8080/api/v1/namespaces/development/services/#{name}"
+    stub_request(:put, expected_url)
+      .to_return(body: open_test_file('service_update.json'), status: 201)
+
+    client = Kubeclient::Client.new 'http://localhost:8080/api/', 'v1'
+    client.update_entity entity, service
```
Comment (abonas) [inline test/test_service.rb]: preferred way to do it is: client.update_service serviceObj
